KAFKA-19136: Move metadata-related configs from KRaftConfigs to MetadataLogConfig

LoganZhuZzz opened the pull request:
Separates metadata-related configurations from the `KRaftConfigs` into
the `MetadataLogConfig` class.

Previously, metadata-related configs were placed in `KRaftConfigs`,
which mixed server-related configs (like process.roles) with
metadata-specific ones (like metadata.log.*), leading to confusion and
tight coupling.

In this PR:
- Extract metadata-related config definitions and variables from
`KRaftConfig` into `MetadataLogConfig`.
- Move `node.id` out of `MetadataLogConfig` into `KafkaMetadataLog’s
constructor` to avoid redundant config references.
- Leave server-related configurations in `KRaftConfig`, consistent with
its role.

This separation makes `KafkaConfig` and `KRaftConfig` cleaner, and
aligns with the goal of having a dedicated MetadataLogConfig class for
managing metadata-specific configurations.

Reviewers: PoAn Yang <payang@apache.org>, Ken Huang
 <s7133700@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

LoganZhuZzz committed:
KAFKA-19136: Move metadata-related configs from KRaftConfigs to MetadataLogConfig

LoganZhuZzz committed:
Revert KafkaConfig back to KafkaConfigs and adjust related variable names

chia7712 commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -16,25 +16,119 @@
+     * @param deleteDelayMillis The amount of time to wait before deleting a file from the filesystem
+     * @param nodeId The node id
+     */
+    public MetadataLogConfig(int logSegmentBytes,
```
could you please add a constructor accepting `AbstractConfig`? Also, please move the getters from `KafkaCOnfig` to this constructor. The getters are listed below.

```
config.metadataLogSegmentBytes,
        config.metadataLogSegmentMinBytes,
        config.metadataLogSegmentMillis,
        config.metadataRetentionBytes,
        config.metadataRetentionMillis,
```

chia7712 commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -16,25 +16,119 @@
+    public final int maxBatchSizeInBytes;
+    public final int maxFetchSizeInBytes;
+    public final long deleteDelayMillis;
+    public final int nodeId;
```
Could you please move it to `KafkaMetadataLog`? that is unrelated to this class

FrankYang0529 commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -16,25 +16,119 @@
+    public MetadataLogConfig(int logSegmentBytes,
+                             int logSegmentMinBytes,
+                             long logSegmentMillis,
+                             long retentionMaxBytes,
+                             long retentionMillis,
```
Can we unify the naming here? Both `METADATA_MAX_RETENTION_BYTES_CONFIG` and `METADATA_MAX_RETENTION_MILLIS_CONFIG` have `MAX`, but only `retentionMaxBytes` uses `Max` here. We can follow old naming like `metadataRetentionBytes` and `metadataRetentionMillis`. Or renaming them like `retentionMaxBytes` and `retentionMaxMillis`.

LoganZhuZzz replied:
Makes sense. I prefer keeping the naming aligned with the config keys for clarity, but renaming the fields would require changes in many files. I’m open either way
just wanted to check if we’re okay with that level of change. cc @chia7712

chia7712 replied:
Maybe we can address that in the follow-up to avoid huge patch

FrankYang0529 commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -16,25 +16,119 @@
+                             long logSegmentMillis,
+                             long retentionMaxBytes,
+                             long retentionMillis,
+                             int maxBatchSizeInBytes,
+                             int maxFetchSizeInBytes,
+                             long deleteDelayMillis,
```
@chia7712, The `maxBatchSizeInBytes`, `maxFetchSizeInBytes`, and `deleteDelayMillis` are constant and only used in `KafkaMetadataLog`. How about replacing them with the constant value directly? So we don't need to maintain them here?

chia7712 replied:
I guess they will be configurable in the future, so maybe `MetadataLogConfig` can initialize them based on the constants for now rather than removing them from `MetadataLogConfig`

LoganZhuZzz replied:
These fields have already been moved to `KafkaMetadataLog`, as they are unrelated to MetadataLogConfig.

LoganZhuZzz committed:
Move related fields to KafkaMetadataLog as constructor fields

LoganZhuZzz committed:
resolve conflicts

chia7712 commented on core/src/main/scala/kafka/raft/RaftManager.scala:
```diff
@@ -231,15 +231,11 @@ class KafkaRaftManager[T](
-        KafkaRaftClient.MAX_FETCH_SIZE_BYTES,
-        ServerLogConfigs.LOG_DELETE_DELAY_MS_DEFAULT,
-        config.metadataNodeIDConfig)
+      config.metadataNodeIDConfig,
```
Maybe we can replace  `metadataNodeIDConfig` by `nodeId`. `metadataNodeIDConfig` is redundant

LoganZhuZzz replied:
Thanks for pointing it out! nodeId already exists, so there’s no need to redefine it.

chia7712 commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -16,25 +16,132 @@
+        this.retentionMillis = retentionMillis;
+    }
+
+    public static MetadataLogConfig fromConfig(AbstractConfig config) {
```
Could you please change it to constructor for consistency?
```java
    public MetadataLogConfig(AbstractConfig config) {
        this.logSegmentBytes = config.getInt(METADATA_LOG_SEGMENT_BYTES_CONFIG);
        this.logSegmentMinBytes = config.getInt(METADATA_LOG_SEGMENT_MIN_BYTES_CONFIG);
        this.logSegmentMillis = config.getInt(METADATA_LOG_SEGMENT_MILLIS_CONFIG);
        this.retentionMaxBytes = config.getInt(METADATA_MAX_RETENTION_BYTES_CONFIG);
        this.retentionMillis = config.getInt(METADATA_MAX_RETENTION_MILLIS_CONFIG);
    }
```

chia7712 commented on core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:
```diff
@@ -581,10 +584,14 @@ object KafkaMetadataLog extends Logging {
     time: Time,
     scheduler: Scheduler,
+    nodeId: Int,
+    maxBatchSizeBytes: Int,
```
`maxBatchSizeBytes`, `maxFetchSizeBytes`, and `logDeleteDelayMs` can be moved to `MetadataLogConfig`.

LoganZhuZzz committed:
Move metadata-related configs

chia7712 commented on core/src/test/scala/kafka/raft/KafkaMetadataLogTest.scala:
```diff
@@ -705,18 +695,7 @@ final class KafkaMetadataLogTest {
-      DefaultMetadataLogConfig.nodeId
-    )
-    val log = buildMetadataLog(tempDir, mockTime, config)
+    val log = buildMetadataLog(tempDir, mockTime, DefaultMetadataLogConfig)
```
```scala
    val config = new MetadataLogConfig(
      DefaultMetadataLogConfig.logSegmentBytes,
      DefaultMetadataLogConfig.logSegmentMinBytes,
      DefaultMetadataLogConfig.logSegmentMillis,
      DefaultMetadataLogConfig.retentionMaxBytes,
      DefaultMetadataLogConfig.retentionMillis,
      maxBatchSizeInBytes,
      DefaultMetadataLogConfig.maxFetchSizeInBytes,
      DefaultMetadataLogConfig.deleteDelayMillis,
    )
```

LoganZhuZzz replied:
I’ve updated it to use the correct value, and the test now passes as expected.

LoganZhuZzz committed:
correct logic in testMaxBatchSize to ensure it passes

m1a2st reviewed (approved):
LGTM

chia7712 commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -16,25 +16,161 @@
+        this.deleteDelayMillis = deleteDelayMillis;
+    }
+
+    public MetadataLogConfig(AbstractConfig config,
```
```java
    public MetadataLogConfig(AbstractConfig config) {
        this(config.getInt(METADATA_LOG_SEGMENT_BYTES_CONFIG),
                config.getInt(METADATA_LOG_SEGMENT_MIN_BYTES_CONFIG),
                config.getLong(METADATA_LOG_SEGMENT_MILLIS_CONFIG),
                config.getLong(METADATA_MAX_RETENTION_BYTES_CONFIG),
                config.getLong(METADATA_MAX_RETENTION_MILLIS_CONFIG),
                KafkaRaftClient.MAX_BATCH_SIZE_BYTES,
                KafkaRaftClient.MAX_FETCH_SIZE_BYTES,
                ServerLogConfigs.LOG_DELETE_DELAY_MS_DEFAULT);
    }
```

Please consider using above style for consistency

LoganZhuZzz replied:
Updated as suggested, thanks!

LoganZhuZzz committed:
Refactor MetadataLogConfig to use default constants for batch size, fetch size, and delete delay

- Removed constructor parameters: maxBatchSizeInBytes, maxFetchSizeInBytes, and deleteDelayMillis
- Replaced them with default values from KafkaRaftClient and ServerLogConfigs
- Simplifies instantiation by relying on standard configuration constants
